liblocation/src/main/java/com/mapbox/android/core/location/LocationCallbackTransport.java

```diff
@@ -13,30 +17,62 @@
  */
 public class LocationCallbackTransport implements LocationListener {
   private static final String TAG = "CallbackTransport";
+  private static final int MESSAGE_WHAT = 0;
   protected final LocationEngineCallback<LocationEngineResult> callback;
   private Location currentBestLocation;
-  private long lastUpdateTime = 0;
   private long fastestInterval = 0;
+  private Location cacheLocation;
+  private Handler handler;
+  private HandlerThread handlerThread;
 
   LocationCallbackTransport(LocationEngineCallback<LocationEngineResult> callback) {
     this.callback = callback;
+    handlerThread = new HandlerThread(this.getClass().getName());
+    handlerThread.start();
+    handler = new ThresholdHandler(handlerThread.getLooper());
+  }
+
+  class ThresholdHandler extends Handler {
+    ThresholdHandler(Looper looper) {
+      super(looper);
+    }
+
+    @Override
+    public void handleMessage(Message msg) {
+      Location location = cacheLocation;
+      cacheLocation = null;
+      if (location != null && sendLocation(location)) {
+        //send location success, start to send next one after fastestInterval period
+        this.sendEmptyMessageDelayed(MESSAGE_WHAT, fastestInterval);
+      }
+    }
   }
 
   @Override
   public void onLocationChanged(Location location) {
-    if (isBetterLocation(location, currentBestLocation)) {
-      currentBestLocation = location;
+    if (fastestInterval != 0) {
+      if (!handler.hasMessages(MESSAGE_WHAT)) {
+        //haven't send message in the previous fastInterval period, so send it directly.
+        sendLocation(location);
+        handler.sendEmptyMessageDelayed(MESSAGE_WHAT, fastestInterval);
+      } else {
+        //cacheLocation will always be the latest one.
+        cacheLocation = location;
```

Comment:
We should cache only the location samples that come close to the interval finish time. Maybe a solution can be posting another message delayed by a fraction of the `fastestInterval` and cache location samples only if that message is not in the queue?